Support `certificate_authorities` extension in ClientHello

s-arash opened the pull request:
This PR adds support for the [`certificate_authorities`](https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.4) extension in ClientHello by:

 - On the client side, adding a default method to `ServerCertVerifier` that asks for CA names to be sent with ClientHello,

 - On the server side, adding the method `certificate_authorities()` to the `ClientHello` type, which is provided to `ResolvesServerCert` when the server needs to pick a cert.

Closes #2235

ctz commented on rustls/tests/common/mod.rs:
```diff
@@ -408,6 +415,7 @@ impl KeyType {
 
     pub fn ca_distinguished_name(&self) -> &'static [u8] {
         match self {
+            // TODO: what's up with the &b"blah"[..] ceremony?
```
This gives each branch of the match the same type. (The type of `b"hello"` is `[u8; 5]`, the type of `&b"hello"[..]` is `&[u8]`).

s-arash replied:
Sure, the type of `b"hello"` is `&[u8; 5]`, but it will be coerced to `&[u8]` without the explicit ceremony.

djc replied:
Feel free to remove the extra ceremony if it is unnecessary!

djc commented on rustls/src/server/hs.rs:
```diff
@@ -398,6 +398,12 @@ impl ExpectClientHello {
             .retain(|scheme| suites::compatible_sigscheme_for_suites(*scheme, &client_suites));
 
+        // We adhere to the TLS 1.2 RFC by not exposing this to the cert resolver if TLS version is 1.2
+        let certificate_authorities_extension = if version != ProtocolVersion::TLSv1_2 {
+            client_hello.certificate_authorities_extension()
+        } else {
+            None
+        };
```
Nit: suggest spelling this

```rust
match version {
    ProtocolVersion::TLSv1_2 => None,
    _ => client_hello.certificate_authorities_extension(),
}
```

djc commented on rustls/tests/common/mod.rs:
```diff
@@ -385,7 +385,7 @@ impl KeyType {
     }
 
-    pub fn get_certified_key(&self) -> Result<Arc<CertifiedKey>, Error> {
+    pub fn get_certified_key_with_raw_pub_key(&self) -> Result<Arc<CertifiedKey>, Error> {
```
Nit: pre-existing issue, but if we're renaming anyway, please get rid of `get_` prefixes for these renamed methods (per [API guidelines](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter)).

djc commented on rustls/tests/common/mod.rs:
```diff
@@ -429,6 +437,13 @@ impl KeyType {
         }
     }
+
+    pub fn ca_cert(&self) -> CertificateDer<'_> {
```
Ordering nit: please move this before `get_chain()`.

djc commented on rustls/tests/common/mod.rs:
```diff
@@ -429,6 +437,13 @@ impl KeyType {
+        self.get_chain()
+            .into_iter()
+            .next_back()
+            .unwrap()
```
Please add a comment on why the `unwrap()` is safe.

djc commented on rustls/tests/server_cert_verifier.rs:
```diff
@@ -152,3 +162,220 @@ fn client_can_override_certificate_verification_and_offer_no_signature_schemes()
+        &self,
+        client_hello: rustls::server::ClientHello<'_>,
+    ) -> Option<Arc<rustls::sign::CertifiedKey>> {
+        if let Some(cas_extension) = client_hello.certificate_authorities() {
```
Nit: use let-else to reduce rightward drift.

djc commented on rustls/tests/server_cert_verifier.rs:
```diff
@@ -152,3 +162,220 @@ fn client_can_override_certificate_verification_and_offer_no_signature_schemes()
+                    .unwrap()
+                    .1;
+                if cas_extension.iter().any(|ca_name| {
+                    x509_parser::x509::X509Name::from_der(ca_name.as_ref())
```
Nit: please import `X509Name` directly.

djc commented on rustls/tests/server_cert_verifier.rs:
```diff
@@ -152,3 +162,220 @@ fn client_can_override_certificate_verification_and_offer_no_signature_schemes()
+    verifier: Arc<dyn ServerCertVerifier>,
+    ca_names: Vec<DistinguishedName>,
+}
+impl ServerCertVerifier for ServerCertVerifierWithCasExt {
```
Nit: empty line before this `impl`, please.

djc commented on rustls/tests/server_cert_verifier.rs:
```diff
@@ -152,3 +162,220 @@ fn client_can_override_certificate_verification_and_offer_no_signature_schemes()
+impl ResolvesServerCert for ResolvesCertChainByCaName {
+    fn resolve(
+        &self,
+        client_hello: rustls::server::ClientHello<'_>,
```
Please import all the types instead of keeping them qualified.

djc commented on rustls/tests/server_cert_verifier.rs:
```diff
@@ -152,3 +162,220 @@ fn client_can_override_certificate_verification_and_offer_no_signature_schemes()
+    }
+}
+
+#[test]
```
Ordering nit: move the tests before any utility types they need.

djc reviewed (commented):
This is looking pretty good to me, just a bunch of nits to address.

s-arash commented:
> This is looking pretty good to me, just a bunch of nits to address.

Thanks @djc. I believed I addressed all the review comments. 

One point: I also removed the `certificate_authorities_ext` field from `ServerConnectionData` (from the earlier version of my PR), as it was not needed, and it brings handling of the certificate authorities extension more in line with the way other Client Hello extensions are handled.

djc commented on rustls/tests/server_cert_verifier.rs:
```diff
@@ -152,3 +163,218 @@ fn client_can_override_certificate_verification_and_offer_no_signature_schemes()
+    ca_names: Vec<DistinguishedName>,
+}
+
+impl ServerCertVerifier for ServerCertVerifierWithCasExt {
+    fn verify_server_cert(
+        &self,
+        end_entity: &pki_types::CertificateDer<'_>,
+        intermediates: &[pki_types::CertificateDer<'_>],
+        server_name: &pki_types::ServerName<'_>,
+        ocsp_response: &[u8],
+        now: pki_types::UnixTime,
+    ) -> Result<rustls::client::danger::ServerCertVerified, Error> {
+        self.verifier
+            .verify_server_cert(end_entity, intermediates, server_name, ocsp_response, now)
+    }
+
+    fn verify_tls12_signature(
+        &self,
+        message: &[u8],
+        cert: &pki_types::CertificateDer<'_>,
+        dss: &rustls::DigitallySignedStruct,
+    ) -> Result<rustls::client::danger::HandshakeSignatureValid, Error> {
+        self.verifier
```
Suggest importing all the qualified types here.

s-arash replied:
Fixed!

djc reviewed (approved):
LGTM, thanks!

ctz reviewed (approved):
Thanks!

(FWIW I went looking if there were some latent tests for this we could enable in bogo, but unfortunately there are not.)

cpu commented on rustls/tests/common/mod.rs:
```diff
@@ -408,25 +408,13 @@ impl KeyType {
-                &b"0\x211\x1f0\x1d\x06\x03U\x04\x03\x0c\x16ponytown ECDSA p521 CA"[..]
-            }
-            KeyType::Ed25519 => &b"0\x1c1\x1a0\x18\x06\x03U\x04\x03\x0c\x11ponytown EdDSA CA"[..],
+            KeyType::Rsa2048 => b"0\x1f1\x1d0\x1b\x06\x03U\x04\x03\x0c\x14ponytown RSA 2048 CA",
```
Thanks for pulling this tidy out into a separate commit :-)

cpu reviewed (approved):
Nice work :rocket: Thanks

cpu commented on rustls/tests/server_cert_verifier.rs:
```diff
@@ -8,11 +8,25 @@ mod common;
 };
-use rustls::{AlertDescription, Error, InvalidMessage};
+use pki_types::{CertificateDer, ServerName};
+use rustls::client::danger::{HandshakeSignatureValid, ServerCertVerified};
+use rustls::client::{danger::ServerCertVerifier, WebPkiServerVerifier};
```
sorry, last minute nit: I think the `danger::ServerCertVerifier` import could get pulled up into L17 alongside `HandshakeSignatureValid` and `ServerCertVerified`.

s-arash replied:
On it!
